stackslib/src/net/api/postblock_proposal.rs

```diff
@@ -0,0 +1,549 @@
+///  that the stacks-node thinks is acceptable.
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct BlockValidateOk {
+    pub block: NakamotoBlock,
+    pub cost: ExecutionCost,
+    pub size: u64,
+}
+
+/// This enum is used for serializing the response to block
+/// proposal validation.
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+#[serde(tag = "Result")]
+pub enum BlockValidateResponse {
+    Ok(BlockValidateOk),
+    Reject(BlockValidateReject),
+}
+
+impl From<Result<BlockValidateOk, BlockValidateReject>> for BlockValidateResponse {
+    fn from(value: Result<BlockValidateOk, BlockValidateReject>) -> Self {
+        match value {
+            Ok(o) => BlockValidateResponse::Ok(o),
+            Err(e) => BlockValidateResponse::Reject(e),
+        }
+    }
+}
+
+impl<T> From<T> for BlockValidateReject
+where
+    T: Into<ChainError>,
+{
+    fn from(value: T) -> Self {
+        let ce: ChainError = value.into();
+        BlockValidateReject {
+            reason: format!("Chainstate Error: {ce}"),
+            reason_code: ValidateRejectCode::ChainstateError,
+        }
+    }
+}
+
+/// Represents a block proposed to the `v2/block_proposal` endpoint for validation
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct NakamotoBlockProposal {
+    /// Proposed block
+    pub block: NakamotoBlock,
+    /// Identifies which chain block is for (Mainnet, Testnet, etc.)
+    pub chain_id: u32,
+}
+
+impl StacksMessageCodec for NakamotoBlockProposal {
+    fn consensus_serialize<W: Write>(&self, fd: &mut W) -> Result<(), CodecError> {
+        write_next(fd, &self.block)?;
+        write_next(fd, &self.chain_id)
+    }
+
+    fn consensus_deserialize<R: Read>(fd: &mut R) -> Result<Self, CodecError> {
+        Ok(Self {
+            block: read_next(fd)?,
+            chain_id: read_next(fd)?,
+        })
+    }
+}
+
+impl NakamotoBlockProposal {
+    fn spawn_validation_thread(
+        self,
+        sortdb: SortitionDB,
+        mut chainstate: StacksChainState,
+        receiver: Box<dyn ProposalCallbackReceiver>,
+    ) -> Result<JoinHandle<()>, std::io::Error> {
+        thread::Builder::new()
+            .name("block-proposal".into())
+            .spawn(move || {
+                let result = self.validate(&sortdb, &mut chainstate);
+                receiver.notify_proposal_result(result);
+            })
+    }
+
+    /// Test this block proposal against the current chain state and
+    /// either accept or reject the proposal
+    ///
+    /// This is done in 3 stages:
+    /// - Static validation of the block, which checks the following:
+    ///   - Block header is well-formed
+    ///   - Transactions are well-formed
+    ///   - Miner signature is valid
+    /// - Validate threshold signature of stackers
+    /// - Validation of transactions by executing them agains current chainstate.
+    ///   This is resource intensive, and therefore done only if previous checks pass
```

Comment:
It's really a shame that we can't easily overlap this with the similar code in `NakamotoChainState` without substantial refactoring.  Perhaps this is just going to have to be a back-burner issue.

Reply:
Yeah, I agree. The code overlap isn't _so_ excessive, though, because most of the work here is already taking place through invocations of `begin_tenure`, etc. There's also a repeated code block here that needs to be removed: the stacker signature check. That cannot be part of the proposal validation (because the proposal validation is occurring to decide whether or not to sign in the first place).